[wanghejin] iP

leawang0 opened the pull request:
# Cheems

> "Let's get some fries by the pier!" – Cheems

Cheems frees your mind of having to remember things you need to do. It's,
- GUI-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download a release.
2. double-click it.
3. add your tasks.
let it manage your tasks for you 😉

And it is **_FREE_**!

Features:

- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

# User Guide
## Run Application
### Use Gradle
Download Gradle from [here](https://gradle.org/install/) and configure it in your choice of IDE. If you wish, you may follow this [tutorial](https://se-education.org/guides/tutorials/gradle.html).
To use Gradle to run the application, run the below command in the terminal
```shell
./gradlew run
```

### Use IDE
To use IDE to run the application, run `main` in `Launcher.java`.

## Modify GUI
You can modify how your graphical interface looks like in `./src/Main/resources/view`.
Change how the dialog box looks by editing `DialogBox.fxml`:
```fxml
<fx:root alignment="TOP_RIGHT" maxHeight="Infinity" maxWidth="1.7976931348623157E308" prefWidth="400.0"
         type="javafx.scene.layout.HBox" xmlns="http://javafx.com/javafx/17" xmlns:fx="http://javafx.com/fxml/1">
    <children>
        <Label fx:id="dialog" text="Label with long text that should wrap to the next line when needed." wrapText="true" />
        <ImageView fx:id="displayPicture" fitHeight="99.0" fitWidth="99.0" pickOnBounds="true" preserveRatio="true" />
    </children>
    <padding>
        <javafx.geometry.Insets bottom="15.0" left="5.0" right="5.0" top="15.0" />
    </padding>
</fx:root>
```
Similarly, you can change how the main window looks like by editing `MainWindow.fxml`.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

leawang0 committed:
Customised own chatbot

leawang0 committed:
Achieved Level-1

leawang0 committed:
Achieved Level-2

leawang0 committed:
Achieved Level-3

leawang0 committed:
Changed database type to ArrayList

leawang0 committed:
Achieved Level-4, basic project layout

leawang0 committed:
Cleaner error handling in parser

leawang0 committed:
Handle null string in parser

leawang0 committed:
Passed UI test

leawang0 committed:
feat: added delete action to database

leawang0 committed:
feat: Added format function to UI

leawang0 committed:
test: added error cases for text-ui-test

leawang0 committed:
fix: Correct subject-verb agreement in chat response

leawang0 committed:
fix: change access modifier of ui and scanner in Cheems

leawang0 committed:
fix: changed Database & UI to contain static methods

leawang0 committed:
feat: update file storage upon deletion of task

leawang0 committed:
feat: update file storage upon deletion of task

leawang0 committed:
Merge branch 'branch-Level-7'

leawang0 committed:
feat: recognise date in format yyyy-mm-dd in ddl, event

leawang0 committed:
Merge branch 'branch-Level-8'

leawang0 committed:
fix: Delete unused deleteTask method

leawang0 committed:
Merge branch 'master' into add-gradle-support

leawang0 committed:
Update build.gradle to run Cheems

leawang0 committed:
doc: Add JavaDoc for non-public classes and methods

leawang0 committed:
feat: Add find with keyword feature

leawang0 committed:
refactor: Move all files to package "cheems"

leawang0 committed:
style: Make naming and layout comply to coding standard

leawang0 committed:
Add throws keyword for explicit exception handling

leawang0 committed:
Fix import statements to comply with coding standard

leawang0 committed:
merge: Resolve conflicts between JavaDoc and Coding Standard, merge to master

leawang0 committed:
merge: Resolve conflicts between master and branch-Level-9

leawang0 committed:
style: Add spacings and line breaks to comply with coding standard

leawang0 committed:
Delete folder created by accident

leawang0 committed:
chore: Modify Parser to be clearer in logic

leawang0 committed:
refactor: Extract create task logic in Tasklist to a new function

leawang0 committed:
refactor: Abstract biz logic in Storage and Tasklist

Delete and mark actions in Storage induce similar actions to overwrite
the database. Combined this logic into one function for updating the
whole database.

Multiple functions in Tasklist requires checking for index out of range.
This is abstracted out as a helper function.

leawang0 committed:
style: Modify header comments to include exceptions

leawang0 committed:
Merge branch 'master' into add-gradle-support

leawang0 committed:
refactor: Convert Parser and Tasklist to use instance methods

leawang0 committed:
refactor: Fix major design problem

leawang0 committed:
chore: Format date in Deadline and Event

leawang0 committed:
test: unit test add method and parser method

leawang0 committed:
refactor: Delete unnecessary interfaces for test mocks

leawang0 committed:
feat: Add GUI to replace text-based UI

leawang0 committed:
Merge branch 'branch-Level-10'

leawang0 committed:
style: Enable checkstyle

ong-wei-hong commented on src/main/java/cheems/Cheems.java:
```diff
@@ -0,0 +1,87 @@
+        cheems.run();
+    }
+
+    public String getWelcomeMsg() {
```
`getWelcomeMsg()` in Cheems.java

ong-wei-hong commented on src/main/java/cheems/functionalities/textUi.java:
```diff
@@ -0,0 +1,56 @@
+    /**
+     * Prints welcome message with logo and greetings.
+     */
+    public void showWelcomeMsg() {
```
`showWelcomeMsg()` in textUi.java

ong-wei-hong reviewed (commented):
Hi Hejin,
I really like the way you have structured your project. It really shows your understanding of what this ip requires.
I have a nit tho. I noticed that there are 2 similar functions in Cheems.java and textUi.java. Is there a reason to have both functions?
Overall, good job for ip so far. Jiayous!
Best,
Wei Hong

McNaBry commented on src/main/java/cheems/Cheems.java:
```diff
@@ -0,0 +1,87 @@
+    public String getResponse(String input) {
+        try {
+            if (!input.equals("bye")) {
+                return parser.parseAndExecute(input);
```
This variable naming is clear and concise 😄

McNaBry commented on src/main/java/cheems/exceptions/EmptyArgumentException.java:
```diff
@@ -0,0 +1,11 @@
+ * Custom exception class to represent an input with empty arguments.
+ * This exception is thrown when an input with empty arguments is entered by the user.
+ */
+public class EmptyArgumentException extends RuntimeException {
```
Nice to see that you split the possible exceptions thrown by the chatbot into more specific classes

McNaBry commented on src/main/java/cheems/functionalities/Storage.java:
```diff
@@ -0,0 +1,101 @@
+            return output;
+        } catch (FileNotFoundException e) {
+            String msg = String.format("I cannot find %s! May be accidental deletion, " +
+                    "try restart cheems.Cheems!", file.getName());
```
I noticed a few instances where you start the next line after the + operator. Would it better if you did it before?

McNaBry commented on src/main/java/cheems/functionalities/Tasklist.java:
```diff
@@ -0,0 +1,200 @@
+ * Represents a task list directly interacts with the user.
+ * Interacts with Storage class as well to retrieve and update information to the text database.
+ */
+public class Tasklist {
```
Clear method naming for this class!

McNaBry reviewed (commented):
Good job 😄

leawang0 committed:
gui: Add tab title and text bubble

leawang0 committed:
gui: Stylize scroll bar

leawang0 committed:
fix: Fix text wrap error

leawang0 committed:
Merge branch 'branch-BetterGui'

leawang0 committed:
style: Update parser code style

Parser code now has arrow-headed layers of nested statements.

Reduce the level of nested statements helps with readability of code.

leawang0 committed:
chore: Delete unused code

leawang0 committed:
Add assertions

There is no assertions in the main code.
Assertions help to find out wrong logic or buggy code.

leawang0 committed:
Merge pull request #1 from wanghejin/branch-CodeQuality

Branch code quality

leawang0 committed:
Merge branch 'master' into branch-Assertions

Resolved merge conflicts between code quality and assertions

leawang0 committed:
Merge branch 'master' into branch-Assertions

leawang0 committed:
Merge pull request #2 from wanghejin/branch-Assertions

Add assertions

leawang0 committed:
feat: Add CI with github actions

There is no CI previously in the project.

CI is needed for the following reasons:
* Automated builds
* Early detection of issues
* Quality assurance

leawang0 committed:
Merge branch 'master' of github.com:wanghejin/ip

leawang0 committed:
feat: Implement fuzzy matching for search optimization

Previously find function only finds tasks that exactly matches with user
input. This is not user-friendly as the user might remember the
description wrongly or simply has typo error.

A fuzzy matching algorithm (Levenstein in this case) is needed to
accommodate such type errors and still find the tasks the user might be
asking for.

leawang0 committed:
fix: Fix return statement error when there is no matching results

leawang0 committed:
config: Set the right configurations for generate fat Jar

leawang0 committed:
Merge branch 'temp'

leawang0 committed:
Add dependency for fuzzywuzzy

leawang0 committed:
Add user guide

leawang0 committed:
Update user guide

leawang0 committed:
style: Enhance code style

leawang0 committed:
test: Add a fail testcase for TaskListTest
